frontend/src/metabase/admin/permissions/utils/graph/partial-updates.ts

```diff
@@ -0,0 +1,97 @@
+import _ from "underscore";
+
+import type { GroupsPermissions } from "metabase-types/api";
+
+type AdvancedPermissions = Record<
+  string,
+  { group_id: number; [key: string]: unknown }[] | undefined
+>;
+
+interface AdvancedPermissionsData {
+  modifiedGroupIds: string[];
+  permissions: AdvancedPermissions;
+}
+
+// utils for dealing with partial graph updates
+
+function getModifiedGroupIdsInGraph(
+  groupIds: string[],
+  originalDataPermissions: GroupsPermissions,
+  newDataPermissions: GroupsPermissions,
+) {
+  // find only the groupIds that have had some kind of modification made in their permissions graph
+  return groupIds.filter(groupId => {
+    return !_.isEqual(
+      newDataPermissions[groupId as unknown as number],
+      originalDataPermissions[groupId as unknown as number],
+    );
+  });
+}
+
+function getModifiedAdvancedPermissions(
+  advancedPermissions: AdvancedPermissions,
+  modifiedGroupIds: string[],
+) {
+  const groupIds = new Set(modifiedGroupIds);
+
+  return Object.fromEntries(
+    Object.entries(advancedPermissions).map(([key, items]) => {
+      const modified = (items || []).filter(item =>
+        groupIds.has(`${item.group_id}`),
+      );
+      return [key, modified];
+    }),
+  );
+}
```

Comment:
I think this and anything else dealing with advanced permissions needs to be in the advanced permissions enterprise plugin

Reply:
Great feedback, I ended up solving this in another way all together. Since impersonation and sandbox state always gets reset on save, we can just send it directly along without picking values off. I refactored the main `get` function in this file to accept an `externallyModifiedGroupIds` so that we can still pick values off of the groups graph when advanced perms are updated, while removing references to advanced permissions in this file.